replay/models/ucb.py

```diff
@@ -0,0 +1,114 @@
+from typing import Optional
+
+import pandas as pd
+
+from pyspark.sql import DataFrame, Window
+from pyspark.sql import functions as sf
+
+from replay.models.base_rec import Recommender
+
+
+class UCB(Recommender):
+    """
+    Calculates upper confidence bound (UCB) for the confidence interval
+    of true fraction of positive ratings.
+
+    ``relevance`` must be converted to binary 0-1 form.
+
+    >>> import pandas as pd
+    >>> data_frame = pd.DataFrame({"user_idx": [1, 2], "item_idx": [1, 2], "relevance": [1, 1]})
+    >>> from replay.utils import convert2spark
+    >>> data_frame = convert2spark(data_frame)
+    >>> model = UCB()
+    >>> model.fit_predict(data_frame,k=1).toPandas()
+       user_idx  item_idx  relevance
+    0         1         2    2.17741
+    1         2         1    2.17741
+
+    """
+    can_predict_cold_users = True
+    can_predict_cold_items = True
+
+    def _fit(
+            self,
+            log: DataFrame,
+            user_features: Optional[DataFrame] = None,
+            item_features: Optional[DataFrame] = None,
+    ) -> None:
+        vals = log.select("relevance").where(
+            (sf.col("relevance") != 1) & (sf.col("relevance") != 0)
+        )
+        if vals.count() > 0:
+            raise ValueError("Relevance values in log must be 0 or 1")
+
+        items_counts = log.groupby("item_idx").agg(
+            sf.sum("relevance").alias("pos"),
+            sf.count("relevance").alias("total"),
+        )
+
+        full_count = log.count()
+        items_counts = items_counts.withColumn(
+            "relevance",
+            (sf.col("pos") / sf.col("total") + sf.sqrt(sf.log(sf.lit(2 * full_count)) / sf.col("total")))
```

Comment:
what is the formula source? I have not found a formula with 2 in numerator, but found one with additional hyperparameter c.
https://towardsdatascience.com/the-upper-confidence-bound-ucb-bandit-algorithm-c05c2bf4c13f
Do we also need it?

Reply:
we can add this parameter. I took formulae from https://russianblogs.com/article/9534740465/

Reply:
in TDS article _c_ comes before sqrt, unlike in your code. _c_ there means "increase the exploration by _c_ times". I believe it is better to place _c_ before exploration term